File: Serenity.Script.UI/Misc/GoogleMap.cs

namespace Serenity
{
    using google.maps;
    using jQueryApi;
    using System;
    using System.Html;
    using System.Runtime.CompilerServices;

    [Imported(ObeysTypeSystem = true), Element("<div/>")]
    public class GoogleMap : Widget<GoogleMapOptions>
    {
        private Map map;

        public GoogleMap(jQueryObject container, GoogleMapOptions opt)
            : base(container, opt)
        {
            var center = new LatLng(options.Latitude ?? 0, options.Longitude ?? 0);

            map = new Map(container[0], new MapOptions          
            {
                Center = center,
                MapTypeId = options.MapTypeId ?? MapTypeId.roadmap,
                Zoom = options.Zoom ?? 15,
                ZoomControl = true
            });

            if (options.MarkerTitle != null)
            {
                new Marker(new MarkerOptions {
                    Position = new LatLng(
                        options.MarkerLatitude ?? options.Latitude ?? 0, 
                        options.MarkerLongitude ?? options.Longitude ?? 0),
                    Map = map,
                    Title = options.MarkerTitle,
                    Animation = Animation.Drop
                });
            }

            LazyLoadHelper.ExecuteOnceWhenShown(container, () =>
            {
                GEvent.Trigger(map, "resize");
                map.SetCenter(center); // in case it wasn't visible (e.g. in dialog)
            });
        }

        public Map Map
        {
            get { return map; }
        }
    }

    [Imported, Serializable]
    public class GoogleMapOptions
    {
        public double? Latitude { get; set; }
        public double? Longitude { get; set; }
        public Double? Zoom { get; set; }
        public MapTypeId? MapTypeId { get; set; }
        public string MarkerTitle { get; set; }
        public double? MarkerLatitude { get; set; }
        public double? MarkerLongitude { get; set; }
    }
}

File: src/Serenity.Scripts/CoreLib/UI/Widgets/GoogleMap.ts
namespace Serenity {
    export interface GoogleMapOptions {
        latitude?: any;
        longitude?: any;
        zoom?: any;
        mapTypeId?: any;
        markerTitle?: string;
        markerLatitude?: any;
        markerLongitude?: any;
    }

    @Decorators.registerEditor('Serenity.GoogleMap', [])
    @Decorators.element('<div/>')
    export class GoogleMap extends Widget<GoogleMapOptions> {

        private map: any;

        constructor(container: JQuery, opt: GoogleMapOptions) {
            super(container, opt);

            // @ts-ignore
            var center = new google.maps.LatLng(
                Q.coalesce(this.options.latitude, 0),
                Q.coalesce(this.options.longitude, 0));

            var mapOpt: any = new Object();
            mapOpt.center = center;
            mapOpt.mapTypeId = Q.coalesce(this.options.mapTypeId, 'roadmap');
            mapOpt.zoom = Q.coalesce(this.options.zoom, 15);
            mapOpt.zoomControl = true;
            // @ts-ignore
            this.map = new google.maps.Map(container[0], mapOpt);

            if (this.options.markerTitle != null) {
                var markerOpt: any = new Object();

                var lat = this.options.markerLatitude;
                if (lat == null) {
                    lat = Q.coalesce(this.options.latitude, 0);
                }
                var lon = this.options.markerLongitude;
                if (lon == null) {
                    lon = Q.coalesce(this.options.longitude, 0);
                }
                // @ts-ignore
                markerOpt.position = new google.maps.LatLng(lat, lon);
                markerOpt.map = this.map;
                markerOpt.title = this.options.markerTitle;
                markerOpt.animation = 2;
                // @ts-ignore
                new google.maps.Marker(markerOpt);
            }

            Serenity.LazyLoadHelper.executeOnceWhenShown(container, () => {
                // @ts-ignore
                google.maps.event.trigger(this.map, 'resize', []);
                this.map.setCenter(center);
                // in case it wasn't visible (e.g. in dialog)
            });
        }

        get_map(): any {
            return this.map;
        }
}
}